### ISSUE TITLE
Duplicate desktop id

### ISSUE THREAD

**Issue by msteen** [baskerville/bspwm#690]
While experimenting with a better desktop setup in my `bspwmrc`, I have restarted `bspwm` (loop example) a bunch of time, and somehow ended up with two desktops that share the same id.

**Comment by msteen** [comment]
I just experienced the issue again and managed to reproduce the problem at will. The problem lies within the `--reset-desktops` command. If you have `--reset-desktops` in your `bspwmrc` name a desktop on monitor 1 and move it to monitor 2 and then execute `bspwmrc` again it will cause the duplicate desktop id problem to occur. So this leads me to believe there is a bug within `--reset-desktops`.

**Comment by msteen** [comment]
After some more debugging I can easily reproduce by having `--reset-desktops` in your `bspwmrc` script and then by removing a desktop and restarting `bspwm` as in the loop example. All the times I triggered the issue was by moving a desktop from one monitor to another, but that has the same effect of removing a desktop from a monitor, it just happens to be added back again to another monitor. Do you have any idea how X generates the id, because I always seem to get the same desktop ids and all you do to generate the id just call `xcb_generate_id(dpy)`, so nothing special. I am quite sure this is the root of the problem, X or XCB generates an id for you using something akin to a seeded pseudo random number generator, so the generated ids are deterministic. This causes duplicate ids by generating an id for the added desktop that is already in use in the state loaded by `bspc wm --load-state`. A few potential fixes would be:

1. Keep generating new ids until you find one that does not already exist.
2. Somehow cause X or XCB to use a different seed for the id generation.

And it seems unlikely that this problem is limited to desktops, instead it probably will cause problems everywhere `bspwm` generate ids, e.g. monitors as well. Maybe look at how `i3` handles this issue, since I thought they also supported the dumping and loading of the window manager state.

**Comment by baskerville** [comment]
Thanks, could you provide a minimal `bspwmrc` and a minimal set of instructions on how to reproduce?

**Comment by msteen** [comment]
Start the window manager as follows, location will dependent on OS:

    while true; do
      bspwm || break
    done

Have the following somewhere in your startup scripts:

    export BSPWM_STATE="$XDG_RUNTIME_DIR/bspwm/state.json"

A minimal `bspwmrc` file that should trigger the issue would be:

    #!/usr/bin/env bash

    mkdir -p $(dirname "$BSPWM_STATE")

    if [ -e "$BSPWM_STATE" ] ; then
      bspc wm --load-state "$BSPWM_STATE"
      rm "$BSPWM_STATE"
    fi

    bspc monitor --reset-desktops 1 2


Once `bspwm` is started you should run the following in a shell:
`bspc desktop 2 --remove`

By the way, is it not a bug that the `--remove` command only works on empty desktops? Should it not just move the nodes to the remaining desktops, unless it is the last remaining one?

Restart `bspwm` by running the following in a shell:
`bspc wm --dump-state > "$BSPWM_STATE" && bspc quit 0`

These steps should allow you to reproduce the bug. The following should return 1 instead of 2:
`bspc query -D | sort | uniq | wc -l`

I use the following to revert back to a valid state:
`pkill bspwm && sleep 0.5 && bspc wm -o`

**Comment by baskerville** [comment]
Thanks, I was able to reproduce. The problem comes from the fact that `restore_tree` doesn't generate any calls to `xcb_generate_id` (that's intentional). This would be fine if we had total control of how IDs are generated. So we might generate the IDs ourselves. But one problem would remain: we don't want the internal nodes IDs to clash with the windows IDs.

I found a terrible platform specific bug while reproducing this bug: on *Darwin* `bspc wm -d > foo` results in `foo` being empty most of the times. This is caused by 98fe29a. I'm hoping that it doesn't happen on *FreeBSD*!
